flent/runners.py

```diff
@@ -950,6 +956,126 @@ def find_binary(cls, host, interval, length, ip_version, local_bind=None,
         raise RuntimeError("No suitable Iperf binary found.")
 
 
+class ParseException(Exception):
+    pass
+
+
+class SsRunner(ProcessRunner):
+    """Runner for iterated `ss -t -i -p`. Depends on same partitial output
+    separationa and time stamping as TcRunner."""
+
+    ip_v4_addr_sub_re = "(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])"
+    # ref.: to commented, untinkered version: ISBN 978-0-596-52068-7
+    ip_v6_addr_sub_re = "(?:(?:[A-F0-9]{1,4}:){7}[A-F0-9]{1,4}|(?=(?:[A-F0-9]{0,4}:){0,7}[A-F0-9]{0,4}(?![:.\w]))(([0-9A-F]{1,4}:){1,7}|:)((:[0-9A-F]{1,4}){1,7}|:))(:\d+)+"
+
+    time_re = re.compile(r"^Time: (?P<timestamp>\d+\.\d+)", re.MULTILINE)
+    cwnd_re = re.compile(r"cwnd:(?P<cwnd>\d+)", re.MULTILINE)
+    rtt_re = re.compile(r"rtt:(?P<rtt>\d+\.\d+/\d+\.\d+)", re.MULTILINE)
+    pid_re = re.compile(r"pid=(?P<pid>\d+)", re.MULTILINE)
+    ports_ipv4_re = re.compile(r"" + ip_v4_addr_sub_re +
+                               ":(?P<src_p>\d+)\s+?" + ip_v4_addr_sub_re + ":(?P<dst_p>\d+)")
+    ports_ipv6_re = re.compile(r"" + "(?P<src_p>" + ip_v6_addr_sub_re + ")" + "\s+" +
+                               "(?P<dst_p>" + ip_v6_addr_sub_re + ")", re.IGNORECASE)
+
+    src_p = []
+    dst_p = []
+    # ref.: upstream ss
+    ss_states = ["UNKNOWN", "ESTAB", "SYN-SENT", "SYN-RECV", "FIN-WAIT-1",
+                 "FIN-WAIT-2", "TIME-WAIT", "UNCONN", "CLOSE-WAIT", "LAST-ACK",
+                 "LISTEN", "CLOSING"]
+    ss_states_re_pat = "|".join(ss_states)
+
+    def __init__(self, **kwargs):
+        super(SsRunner, self).__init__(**kwargs)
+        self.pid = str(self._parent.pid)
+
+    def filter_np_parent(self, part):
+        sub_part = []
+        sub_parts = re.split(self.ss_states_re_pat, part)
+        sub_parts = [sp for sp in sub_parts if sp.strip()]
+
+        for sp in sub_parts:
+            pid = self.pid_re.search(part)
+            if None is pid:
+                continue
+            pid_str = pid.group('pid')
+            if self.pid == pid_str:
+                sub_part.append(sp)
+
+        if 1 != len(sub_part):
+            raise ParseException()
+
+        return sub_part[0]
+
+    def parse_part(self, part):
+
+        sub_part = self.filter_np_parent(part)
+
+        timestamp = self.time_re.search(sub_part)
+        cwnd = self.cwnd_re.search(sub_part)
+        rtt = self.rtt_re.search(sub_part)
```

Comment:
Just skipping the entry if the regexes don't match (as you do now) is fine. Empty results when an error happens is OK. Crashing, not so much ;)

As for unittests, there's a single one for the Cake parser in unittests/test_parsers.py - because that kept breaking. Have not added anything else so far; but feel free to add one for this if you want :)